pkg/scheduler/allocator.go

```diff
@@ -214,6 +211,56 @@ func (m *Scheduler) tryBatchAllocation(partition string, candidates []*Schedulin
     return allocations, failedAsks
 }
 
+// TODO: convert this as an interface.
+func evaluateForSchedulingPolicy(m *Scheduler, nodes []*SchedulingNode, partition string,
+    candidate *SchedulingAllocationAsk, partitionContext *PartitionSchedulingContext) []*SchedulingNode {
+
+    // If bin-packing is not enabled, simply return
+    if !partitionContext.partition.NeedBinPackingSchedulingPolicy() {
+        // Sort by MAX_AVAILABLE resources.
+        // TODO, this should be configurable.
+        SortNodes(nodes, MaxAvailableResources)
+        return nodes
+    }
+
+    // Do an in-place sorting
+    nodes = m.SortAllNodesWithAscendingResource(partition)
+
+    for i := 0; i < len(nodes); i++ {
+        node := nodes[i]
+        if !node.CheckAllocateConditions(candidate.AskProto.AllocationKey) {
+            // skip the node if conditions can not be satisfied
+            continue
+        }
+
+        // once we have a node which can fit the allocation ask, send back for scheduling
+        if node.CheckAndAllocateResource(candidate.AllocatedResource, false /* preemptionPhase */) {
+            if log.IsDebugEnabled() {
+                log.Logger().Debug("Selected node:",
+                    zap.Any("id", node))
+            }
+            return append(make([]*SchedulingNode, 0, 1), node)
+        }
+    }
+
+    return nodes
+}
+
+func getNodeList(m *Scheduler, partition string) []*SchedulingNode {
+    nodeList := m.clusterInfo.GetPartition(partition).CopyNodeInfos()
+    if len(nodeList) <= 0 {
+        // When we don't have node, do nothing
+        return nil
+    }
+
+    schedulingNodeList := make([]*SchedulingNode, len(nodeList))
+    for idx, v := range nodeList {
+        schedulingNodeList[idx] = NewSchedulingNode(v)
+    }
+
+    return schedulingNodeList
+}
+
```

Comment:
The `getNodeList` is the old logic, I think we should try to avoid a full copy of node list and creation of scheduling node after that.

Reply:
iterators are implemented and passed around